flask_security/decorators.py

```diff
@@ -53,7 +53,11 @@ def _get_unauthorized_view():
             except BuildError:
                 view = None
         utils.do_flash(*utils.get_message('UNAUTHORIZED'))
-        return redirect(view or request.referrer or '/')
+        redirect_to = '/'
+        if request.referrer and not request.referrer.endswith(request.path):
```

Comment:
I hope it will be clearer from the code:
```diff
diff --git a/flask_security/decorators.py b/flask_security/decorators.py
index e1799a2..c1f00f0 100644
--- a/flask_security/decorators.py
+++ b/flask_security/decorators.py
@@ -54,7 +54,8 @@ def _get_unauthorized_view():
                 view = None
         utils.do_flash(*utils.get_message('UNAUTHORIZED'))
         redirect_to = '/'
-        if request.referrer and not request.referrer.endswith(request.path):
+        if request.referrer and not request.referrer.endswith(
+                request.full_path.rstrip('?')):
             redirect_to = request.referrer

         return redirect(view or redirect_to)
diff --git a/tests/test_common.py b/tests/test_common.py
index 4e76a33..2111dd3 100644
--- a/tests/test_common.py
+++ b/tests/test_common.py
@@ -136,12 +136,12 @@ def test_unauthorized_access(client, get_message):
 def test_unauthorized_access_with_referrer(client, get_message):
     authenticate(client, 'joe@lp.com')
     response = client.get('/admin', headers={'referer': '/admin'})
-    assert response.headers['Location'] != '/admin'
+    assert response.headers['Location'][-6:] != '/admin'
     client.get(response.headers['Location'])

     response = client.get('/admin',
                           headers={'referer': 'http://localhost/admin'})
-    assert response.headers['Location'] != '/admin'
+    assert response.headers['Location'][-6:] != '/admin'
     client.get(response.headers['Location'])

     response = client.get('/admin',
@@ -149,6 +149,14 @@ def test_unauthorized_access_with_referrer(client, get_message):
                           follow_redirects=True)
     assert response.data.count(get_message('UNAUTHORIZED')) == 1

+    response = client.get('/admin?x=1',
+                          headers={'referer': '/admin?x=1'},
+                          follow_redirects=True)
+    assert response.data.count(get_message('UNAUTHORIZED')) == 1
+
+    response = client.get('/admin?x=1',
+                          headers={'referer': '/admin?x=0'})
+    assert response.headers['Location'][-10:] == '/admin?x=0'

 def test_roles_accepted(client):
     for user in ("matt@lp.com", "joe@lp.com"):
```